dali/kernels/imgproc/resample/resampling_impl_cpu.h

```diff
@@ -83,6 +85,148 @@ void ResampleCol(Out *out, const In *in, int x, int w, const int32_t *in_columns
   }
 }
 
+template <typename Out, typename In>
+struct SIMD_vert_resample_impl {
+#ifdef __SSE2__
+  static constexpr int kVecSize = 16;
+  static constexpr int load_lanes = kVecSize / sizeof(In);
+  static constexpr int store_lanes = kVecSize / sizeof(Out);
+  static constexpr int kNumLanes = load_lanes > store_lanes ? load_lanes : store_lanes;
+  static constexpr int kNumVecs = kNumLanes * sizeof(float) / kVecSize;
+
+  using vec_pack = simd::multivec<kNumVecs>;
+#endif
+
+  static void run(Out *out, const In **rows, const float *kernel, int support,
+                   int begin_col, int end_col) {
+    int i = begin_col;
+#ifdef __SSE2__
+    for (; i + kNumLanes <= end_col; i += kNumLanes) {
+      vec_pack vtmp = vec_pack::zero();
+
+      for (int k = 0; k < support; k++) {
+        vec_pack vin = vec_pack::load(rows[k] + i);
+        __m128 coeff = _mm_set1_ps(kernel[k]);
+        for (int v = 0; v < kNumVecs; v++)
+          vtmp.v[v] = _mm_add_ps(vtmp.v[v], _mm_mul_ps(coeff, vin.v[v]));
+      }
+      store(out + i, vtmp);
+    }
+#endif
+    for (; i < end_col; i++) {
+      float tmp = 0;
+      for (int k = 0; k < support; k++)
+        tmp += rows[k][i] * kernel[k];
+      out[i] = ConvertSat<Out>(tmp);
+    }
+  }
+};
+
+
+template <typename Out, typename In>
+struct SIMD_horz_resample_impl {
+#ifdef __SSE2__
+  static constexpr int kVecSize = 16;
+  static constexpr int kNumLanes = kVecSize / sizeof(Out);
+  static constexpr int kNumVecs = kNumLanes * sizeof(float) / kVecSize;
+
+  using vec_pack = simd::multivec<kNumVecs>;
+#endif
+
+  template <int static_channels, bool clamp_left, bool clamp_right>
+  inline int run(Out *out, const In *in, int ox0, int ox1, int w,
+                 const int32_t *in_columns,
+                 const float *coeffs, int support,
+                 int dynamic_channels) {
+    const int channels = static_channels < 0 ? dynamic_channels : static_channels;
+
+    int x = ox0;
+#ifdef __SSE2__
+    float tmpin[kNumLanes];
+    for (; x + kNumLanes <= ox1; x += kNumLanes) {
+      Out tmp_out[kNumLanes];
+
+      if (static_channels < 0) {
+        // we don't know how many channels we have at compile time - inner loop over filter
+        for (int c = 0; c < channels; c++) {
```

Comment:
What difference it would make if we have a variants for predefined number of channels - 1 and 3 for example?

Reply:
Then it goes for the branch below (static_channels) and then `channels` is effectively a constant expression. It's quite optimized. We _could_ use a specialized variant for 1 channel, 2, and 4 channels - 3 not so much, interleaving and deinterleaving triples on X86 is a pain.
We could use it on ARM NEON, though, with its beautiful interleaved load and store.

Reply:
So when the number of channels is not static?

Reply:
I guess it's for some weird cases not covered by static switch for the number of channels, like 7 or 16.